File: dahlia/src/main/kotlin/com/github/saturn_siv/palm/dahlia/plugins/cms/models/Tag.kt
package com.github.saturn_siv.palm.dahlia.plugins.cms.models

import jakarta.persistence.*
import java.time.Instant

@Table(name = "tags")
@Entity(name = "tag")
class Tag(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(nullable = false)
    var id: Int,
    @Column(nullable = false)
    var name: String,
    @Column(nullable = false)
    var version: Int,
    @Column(nullable = false)
    var updatedAt: Instant,
    @Column(nullable = false)
    var createdAt: Instant,

    ) {
    @OneToMany(cascade = [(CascadeType.ALL)], fetch = FetchType.LAZY, mappedBy = "tag")
    var resources = mutableSetOf<TagResource>()
}

File: daisy/models/tag.go
package models

import (
	"time"
)

type Tag struct {
	Model

	Name string `gorm:"uniqueIndex;not null;size:63"`

	Resources []*TagResource
}

func (Tag) TableName() string {
	return "tags"
}

type TagResource struct {
	ID           uint   `gorm:"primarykey"`
	TagID        uint   `gorm:"not null"`
	ResourceType string `gorm:"index;not null;size:127"`
	ResourceId   uint
	CreatedAt    time.Time `gorm:"not null"`

	Tag *Tag
}

func (TagResource) TableName() string {
	return "tags_resources"
}
